compiler/rustc_infer/src/infer/error_reporting/mod.rs

```diff
@@ -2172,6 +2172,23 @@ impl<'a, 'tcx> InferCtxt<'a, 'tcx> {
                                 }
                             }
                         }
+                        (expected_type, ty::Adt(found_type, _)) => {
+                            let is_found_result = self.tcx.is_diagnostic_item(sym::Result, found_type.did());
+                            let is_expected_result = match expected_type {
+                                ty::Adt(idk, _) => self.tcx.is_diagnostic_item(sym::Result, idk.did()),
+                                _ => false
+                            };
+                            if is_found_result && !is_expected_result {
+                                if let Ok(code) = self.tcx.sess().source_map().span_to_snippet(span) {
+                                    err.span_suggestion(
+                                        span,
+                                        "Either change the return type or use `unwrap`",
+                                        format!("{}.unwrap()", code),
```

Comment:
I don't think we should ever suggest `.unwrap()`, even if that fixes the code. How about linking to the rust book  chapter about error handling?

Reply:
Would you please suggest some specific wording?  This is my first contribution so would appreciate the guidance :)

Reply:
follow-up:

As I said in the PR description, it would be nice if in the case of 
```rs
fn parse(num: String) -> i32 {
    num.parse()
}
```
I could suggest changing the expression type to be `Result<i32, std::num::ParseIntError>` instead of i32.  Getting `i32` is easy, but I'm not sure how to get ` std::num::ParseIntError`

Reply:
Hmm, I'm not sure about better wording here, but the book is mentioned in other error messages, maybe you'd find some inspiration by searching for references to the book in the compiler.

Reply:
As for suggesting to change the return type, I'm not sure what the best way to get it would be here. I think you need the function signature for that. I don't have time to help you further though, sorry.

Reply:
Thank you.  I'll take a look at the book.

What about the second question I asked under "Questions for reviewer"?  This is my first contribution, so if you could recommend a reviewer potentially capable of more hand-holding I'd appreciate it 🙂

Reply:
You can write `r? diagnostics` (I think that's the right name?) in a new comment to request a wg-diagnostics member for review, which is the most fitting for this PR since it's a diagnostic change.